src/main/java/net/neoforged/neoforge/registries/DeferredHolder.java

```diff
@@ -0,0 +1,270 @@
+        this.bind(false);
+    }
+
+    /**
+     * Gets the object stored by this DeferredHolder, if this holder {@linkplain #isPresent() is present}.<br>
+     *
+     * @throws IllegalStateException If the backing registry is unavailable.
+     * @throws NullPointerException  If the underlying Holder has not been populated (the target object is not registered).
+     */
+    @SuppressWarnings("unchecked")
+    @Override
+    public T value() {
+        bind(true);
+        Objects.requireNonNull(this.holder, () -> "Trying to access unbound value: " + this.key);
+        return (T) this.holder.value();
+    }
+
+    /**
+     * Gets the object stored by this DeferredHolder, if this holder {@linkplain #isPresent() is present}.<br>
+     *
+     * @throws IllegalStateException If the backing registry is unavailable.
+     * @throws NullPointerException  If the underlying Holder has not been populated (the target object is not registered).
+     */
+    @Override
+    public T get() {
+        return this.value();
+    }
+
+    /**
+     * @return An optional containing the target object, if {@link #isPresent()}, otherwise {@linkplain Optional#empty() an empty optional}.
+     */
+    public Optional<T> asOptional() {
+        return isPresent() ? Optional.of(value()) : Optional.empty();
+    }
+
+    /**
+     * The type of the registry is really <? super T> but this saves us some additional ugly casting and doesn't break anything.
+     *
+     * @return The registry that this DeferredHolder is pointing at, or null if it doesn't exist.
+     */
+    @Nullable
+    @SuppressWarnings("unchecked")
+    protected Registry<R> getRegistry() {
+        return (Registry<R>) BuiltInRegistries.REGISTRY.get(this.key.registry());
+    }
+
+    /**
+     * Binds this DeferredHolder to the underlying registry and target object.<br> Has no effect if already bound.
+     *
+     * @param throwOnMissingRegistry If true, an exception will be thrown if the registry is absent.
+     * @throws IllegalStateException If throwOnMissingRegistry is true and the backing registry is unavailable.
+     */
+    protected void bind(boolean throwOnMissingRegistry) {
+        if (this.holder != null) return;
+
+        Registry<R> registry = getRegistry();
+        if (registry != null) {
+            this.holder = registry.getHolder(this.key).orElse(null);
+        } else if (throwOnMissingRegistry) {
+            throw new IllegalStateException("Registry not present for " + this + ": " + this.key.registry());
+        }
+    }
+
+    /**
+     * @return The ID of the object pointed to by this DeferredHolder.
+     */
+    public ResourceLocation getId() {
+        return this.key.location();
+    }
+
+    /**
+     * @return The ResourceKey of the object pointed to by this DeferredHolder.
+     */
+    public ResourceKey<R> getKey() {
+        return this.key;
+    }
+
+    /**
+     * Gets the inner holder. Only exists for encapsulation reasons.
```

Comment:
If this is supposed to be package-private you can make it so. (Not sure)

Reply:
no this exists for future expansion of DR and DH to have special variants for stuff like items (where DH then would implement ItemLike)

Reply:
Why would that require the inner holder?

Reply:
Idk what people will need better give them read access than have them ask for it later

Reply:
IMO we should give less rather than more now, especially if we can't come up with a use case. Anything we give now will be hard to take back...

`Only exists for encapsulation reasons.` is unclear to me, maybe just remove it then.

Reply:
DR is currently not extensible anyway as everything one might want to touch, including the constructor, is private.

Reply:
This method was not in the original reference for DH, it doesn't need to exist.  DH *is* `Holder<R>`, so there is no reason to expose the inner holder.